### DISCUSSION TITLE
[HELP] - How to pass multineat.Genome weights to CMA-ES or other optimization algorithms

### DISCUSSION THREAD

**Discussion by itsron143** [ci-group/revolve2#568] [Ask for help]
Hello! 

I have a few questions about implementing a learner in the modular robot evolution cycle for evolving the brain and body (evolution + learning). The learner would be used to optimize the brain parameters for a body and then continue with the evolution of the robot (brain + body)

I'm unable to understand how to pass the weights of `multineat.Genome` of the brain as parameters to CMA-ES.

Following example 4e in revolve2/examples, I was able to understand how to get the weights from `cpg_network_structure` as parameters to CMA-ES and then convert it back to a `BrainCpgNetworkStatic` brain using `uniform_from_params()` method. 

How would the weights of `BrainCpgNetworkNeighborV1` be extracted from `multineat.Genome` to pass it as parameters to CMA-ES and also then develop the `BrainCpgNetworkNeighborV1` brain using the updated parameters? 

Any help would be appreciated!

**Comment by oliverweissl** [comment]
Hello, maybe I don’t understand properly but optimizing the genome is not useful since its just a indirect encoding to the body or brain. Therefore optimization probably wont yield anything. Usually you want to optimize the parameters in the developed brain not in its genome.

In 4e we use this exact setup where the brain is developed, and then its weights are extracted and optimized. As such i think this is also what you want.

Alternatively you can look at the brain class you are using, since all have a `_weight_matrix` attribute you can use this for extracting weights.

**Reply by itsron143** [reply]
thank you for the response! 

(For Lamarckian evolution) When the parameters in the developed brain are optimized, how can they be passed on to the children's brain's genotype? I'd like to pass on the _weight_matrix to the children but the children only store the genotype information.

In 4e, the body is fixed/static and the brain's weights are extracted from the cpg_network_structure. However, in the evolution + learning cycle, in each evolution step the body is also being evolved, so when I pass the optimized brain weights to the children, the brain is unable to develop as the length of the brain weights does not match `cpg_structure.num_connections`

I don't think I fully understand how a CPG brain creates weights from a CPPNWIN network. I'm reading up more about it

again, any help would be appreciated! :)

**Reply by oliverweissl** [reply]
The CPG network is created from the robots hinges (is sensitive to phenotype) then each location in the brain gets queried in the CPPN and therefore gets assigned a weight. (internal + external)

Since your approach bodies evolve you probably wont be able to directly copy the weight matrices into the newly developed brain but the basic approach is:
Parent learns / Parent reproduces -> Child gets developed -> copy learned brain into childs random brain.

However you have to be creative to allow for different brain shapes as your bodies are not guaranteed to be the same. (Maybe only copy parts that fit and query others??)

I hope this helps, btw in the CI Group some people already did lamarcian evolution similar to this so maybe they can give better intuition.
